Add Better Errors for Comparison Chain

hinthornw opened the pull request:
+ change to ABC - this lets us add things like the evaluation name for loading

hinthornw committed:
Add Better Errors for Comparison Chain

agola11 commented:
looks like this needs to be stacked ontop of https://github.com/hwchase17/langchain/pull/6942

hinthornw commented:
> looks like this needs to be stacked ontop of #6942

My fault - should be other way around actually (#6942 stacks on this one)

hinthornw committed:
Add better warnings when ref or inputs are passed but not used (#7115)

If a string evaluator doesn't want references, it currently silently
ignores any that are passed in. This PR proposes to warn once if they're
provided. It does so via a mixin.

Other options we could do:
- Separate classes for criteria with references and comparisons with
references - I don't love this because all the combinations lead to
class explosion and the class names themselves are ugly
(CriteriaWithReferencesNoLabels, etc.)
- Raise errors if they're provided - I don't like this because it
requires more checks when running evaluators over a dataset, but would
be more explicit.
- Decide at calling time -> could be a lot of partial prompt insertions
that makes it more brittle. Would make a single "run" potentially have
two results that are fundamentally different evaluations
